hashicorp/consul
# [CC-7042] Update and enable the HCP metrics sink in the HCP manager

### Description
This PR adds support to initialize the HCP metrics sink without an HCP configuration and to then later update the sink with the HCP config at runtime in the HCP manager. The purpose of this change is to eventually support initiating the linking process via an API request, where that API request will start the HCP manager.

This PR is best viewed commit-by-commit. The general strategy is to add update methods to the HCP telemetry config provider and then call those update methods from the HCP manager.

Changes so that the HCP Telemetry Provider can fetch the config from HCP:
1. Remove the requirement to set the HCP client on the HCP telemetry provider and handle the case when it’s unset.
2. Save the telemetry provider as a dependency and pass it to the HCP manager.
3. In the HCP manager, update the telemetry provider with the HCP client.

Changes so that the metrics client can export to HCP:
1. Add a new interface to the HCP telemetry provider that abstracts the HTTP retry client and headers.
2. Move the setup of the HTTP retry client and headers from the metrics client to the HCP telemetry provider.
3. Allow the telemetry provider’s retry client and headers to be updated with a new HCP config.
4. Pass the HCP configuration to the HCP manager.
5. In the HCP manager, update the telemetry provider with the HCP config.

### Testing & Reproduction steps
Updated relevant unit tests and manually e2e tested this using the following steps:

1. Create a cluster in HCP Consul Central.
2. Configure and start Consul with the cloud configuration provided by HCP Consul Central.
3. Inspect that the telemetry export requests are succeeding in the Consul logs.
4. Inspect the Observability dashboard in HCP Consul Central for the cluster and verify that all the metrics are displayed as expected.

```
2023-12-21T13:22:21.801-0600 [DEBUG] agent.hcp_manager: HCP manager starting
2023-12-21T13:22:21.801-0600 [DEBUG] agent.hcp_manager: updated telemetry config provider with HCP client
2023-12-21T13:22:21.801-0600 [DEBUG] agent.hcp_manager: updated telemetry config provider with HCP configuration
…
2023-12-21T13:24:21.683-0600 [DEBUG] agent.hcp.hcp_telemetry_client: performing request: method=POST url=https://consul-telemetry.cloud.hashicorp.com/otlp/v1/metrics
…
2023-12-21T13:25:21.682-0600 [DEBUG] agent.hcp.hcp_telemetry_client: performing request: method=POST url=https://consul-telemetry.cloud.hashicorp.com/otlp/v1/metrics
…
2023-12-21T13:26:21.699-0600 [DEBUG] agent.hcp.hcp_telemetry_client: performing request: method=POST url=https://consul-telemetry.cloud.hashicorp.com/otlp/v1/metrics
```

### Links
- Jira: https://hashicorp.atlassian.net/browse/CC-7042

### PR Checklist
